### ISSUE TITLE
Upgrade to Helm 3

### ISSUE THREAD

**Issue by rossf7** [giantswarm/roadmap#30]
**User Story:** As a GS team member I want to be able to use Helm 3. To benefit from the improved security model and to no longer have to manage Tiller which takes time and adds complexity.

## References
* Spec: https://intranet.giantswarm.io/docs/architecture-specs-adrs/specs/helm-3-upgrade/
* [Draft Release Notes](https://www.dropbox.com/scl/fi/rcle9tj1oj54netb5sqkk/Helm-3-Release-Notes.paper?dl=0&rlkey=nq013qe7hn8r0lh7iasys1mm1)

## Epics

- [x] Upgrade control planes to Helm 3 https://github.com/giantswarm/giantswarm/issues/10973 WIP @cokiengchiara @rossf7 @tomahawk28 
-
Upgrade tenant clusters to Helm 3 https://github.com/giantswarm/giantswarm/issues/11748 (blocked until after CP upgrade)

**Releases**
- [x] [Azure v12.1.0](https://github.com/giantswarm/releases/tree/master/azure/v12.1.0)
- [x] [AWS v12.3.0](https://github.com/giantswarm/releases/tree/master/aws/v12.3.0)
- [ ] KVM

**Comment by rossf7** [comment]
## Status 6 March

No progress due to PM week. 

Next step is to merge https://github.com/giantswarm/chart-operator/pull/358 to remove the legacy chartconfig controller and then create a helm3 branch to use helm3 branch of helmclient.

**Comment by rossf7** [comment]
## Status 18th March

- helm3 branch of chart-operator now uses helm3 branch of helmclient.
- Work on the data migration from chart to app CRs is in progress.

## TODO

- Complete data migration and have integration test coverage for it.
- app-operator changes and test chart-operator in tenant clusters.
- test chart-operator in control plane clusters.

@cokiengchiara @tomahawk28 I updated the checklist with what I think is the current state.

**Comment by rossf7** [comment]
## Status 23rd March

- @tomahawk28 Worked on data migration from release configmaps to secrets. Using the Helm 2to3 plugin triggered via a k8s job.
- I worked on app-operator getting it working with latest apiextensions and removing e2e-harness since it also uses helmclient.

## TODO

- Merge Helm 3 changes into new helm3 branch in app-operator.
- Complete data migration changes in chart-operator.
- Start integration testing with default apps in tenant clusters since this is less disruptive than testing in control planes.

Progress will be limited due to it being hackathon week.

**Comment by rossf7** [comment]
## Status 30th March

- app-operator changes in helm3 branch complete and ready for testing.
- Data migration process in chart-operator also complete and ready for testing.

## TODO

- Create chart-operator `1.0.0-rc1` release for testing.
- Test chart-operator in tenant clusters first and then in control planes.
- architect-orb and chart changes for Helm3.
  - crd-install hook replaced by crds directory
  - Release.ReleaseTime removed so we need to use Release.Revision

**Comment by cokiengchiara** [comment]
@rossf7 @tomahawk28 I took a stab at Helm 3 Release Notes https://www.dropbox.com/scl/fi/rcle9tj1oj54netb5sqkk/Helm-3-Release-Notes.paper?dl=0&rlkey=nq013qe7hn8r0lh7iasys1mm1 pls take a look and see if it's complete bullshit? :D

**Comment by rossf7** [comment]
@cokiengchiara Nice it looks good! I left some comments and added a few things.

**Comment by cokiengchiara** [comment]
> @cokiengchiara Nice it looks good! I left some comments and added a few things.

thanks @rossf7 !  I have two comments left but it's fine. In terms of process. I think we can leave the draft there, when you make the PR for the release and release notes, you can just use it. So I will do nothing further about it until then. WDYT?

**Comment by rossf7** [comment]
## Status 6th April

- Release notes for platform release were prepared.
- Testing in tenant clusters continued. Upgrade process works except for deleting tiller configmaps and deployment.
- We identified that we need to move releasemigration logic from chart-operator to app-operator to have more control of the upgrade process.
- We also identified that we need to flatten app-operator and chart-operator so we can migrate app CRs in control plane clusters.

## TODO

- Flatten app-operator and chart-operator.
- Fix issues with releasemigration in tenant clusters.
- Start testing upgrade in control planes.

**Comment by cokiengchiara** [comment]
Tested few cases. Looking at errors to see main reason why it gets stuck.

**Comment by rossf7** [comment]
## Status 17th April

- chart-operator flattening done and will be deployed on Tuesday
- releasemigration resource moved from chart-operator to app-operator and we cordon chart CRs to prevent failed helm releases blocking the migration

## TODO

- Flatten app-operator
- Test Helm 2 to 3 migration in tenants and control planes. It's easier to test in tenants as they can be deleted. So we test their first.
- Change release.Time to release.Revision so charts are Helm 3 compatible. We will need to replace existing tarballs.

**Comment by rossf7** [comment]
## Status May 11th

- Fixes got made to the migration logic and we tested upgrading tenant and control plane on Azure.
- Changes to use Release.Time instead of Release.Revision are now done.

## TODO

- Test control plane upgrade on AWS and KVM and plan upgrade for control plane clusters.
- We need a version of app-operator using helm 2 for bootstrapping chart-operator in older releases.
- opsctl ensure appcatalogs needs updating to use Helm 3 and we need to migrate the helm releases.
- Prepare PRs to update helmclient to Helm 3 in other components.
- Prepare platform releases using chart-operator 1.0.0 but this will likely not happen till next week.

**Comment by cokiengchiara** [comment]
After this: **We need a version of app-operator using helm 2 for bootstrapping chart-operator in older releases.**

- [ ] news-ops. We need a window of a few hours to do a high risk change for upgrading Helm 3, at some point. @rossf7 please draft: What is happening? What is potential impact? To whom?
- [ ] Chiara will also ping other POs to share w their teams. That this is happening.

**Comment by cokiengchiara** [comment]
Draft of customer note for Helm 3 upgrade: https://www.dropbox.com/scl/fi/xif71gjxo4o8y7xyq8esu/Customer-Note-for-Helm-3-Upgrade.paper?dl=0&rlkey=pr8v3r9e3iclkt5ma87kz8m9n

@rossf7 @tomahawk28 can you clarify please some open questions in the doc?
@giantswarm/se we would like to give a heads up to customers in support channel regarding helm 3 upgrade. Can you check the doc please if you are fine with the message?

**Comment by yasn77** [comment]
@cokiengchiara @rossf7 Since this is a pretty big change, as a suggestion could we limit the blast radius by selecting a couple of customers and only upgrading them? So initially we only announce to them and do the upgrades/testing. 

If that works fine and depending on the experience we can either decide to roll it out everywhere else or select another couple fo customers and only roll out to them. 

Basically stagger the release in a more controlled way

**Comment by tomahawk28** [comment]
@yasn77 For that, we need to deploy default operators in a few installations as a choice. I doubt we can keep it as a deployment process for long.
If blasting effects from downtime of app-operator worry the customer, we can keep both *-app-collection helm release as the previous version just before helm 3 release, and make the latest change when they agreed on timing. But we still would have a discrepancy between other CPs.

**Comment by rossf7** [comment]
> Since this is a pretty big change, as a suggestion could we limit the blast radius by selecting a couple of customers and only upgrading them? 

@yasn77 Like @tomahawk28 says this isn't really possible with our current deployment pipeline. The app CRs are managed via the app collections which are deployed to all control planes.

There is a collection per provider. So we could do the upgrade per provider as long as nobody else creates releases of app-operator or chart-operator.

The other idea I had was to first upgrade the harder control planes first. Like China and the larger onprem installations. I don't expect us to need to rollback but it would be good to find problems sooner in the process.

WDYT?

**Comment by omegas27** [comment]
Personally, I am not really a big fan of saying to the customers that they cannot upgrade for a couple of hours on such a short notice. Those should probably be brought to sig-project for discussions so everyone is aware.
Should we create a maintenance window calendar so we can know in advance about such cases and maybe share that calendar with customers so they can plan ahead?

**Comment by yasn77** [comment]
Thansk for the explanation @rossf7 @tomahawk28 

That's a bit of a bummer imho :/

> The other idea I had was to first upgrade the harder control planes first. Like China and the larger onprem installations. I don't expect us to need to rollback but it would be good to find problems sooner in the process.
> 
> WDYT?

That could be an option... Focus on one CP that is large enough to give valid feedback then move on to others with things learned. 

Whichever CP that is though, I agree with @omegas27 that fair warning needs to be given and they are fully aware of what is happening.

**Comment by cokiengchiara** [comment]
So, we are worried about three things:
1. Are we aware of how much risk this change entails? And do we feel confident of our plan to mitigate these risk? 
2. We want to make customer fully aware of what is happening. What would "full awareness" entail? [Customer note](https://www.dropbox.com/scl/fi/xif71gjxo4o8y7xyq8esu/Customer-Note-for-Helm-3-Upgrade.paper?dl=0&rlkey=pr8v3r9e3iclkt5ma87kz8m9n) is not enough?
3. Giving enough notice to customers that they cannot upgrade, so that they have enough time to prepare. How much is enough time? 3 days? 1 week? 2 weeks? 1 month?

**Comment by yasn77** [comment]
> So, we are worried about three things:
> 
> 1. Are we aware of how much risk this change entails? And do we feel confident of our plan to mitigate these risk?
> 2. We want to make customer fully aware of what is happening. What would "full awareness" entail? [Customer note](https://www.dropbox.com/scl/fi/xif71gjxo4o8y7xyq8esu/Customer-Note-for-Helm-3-Upgrade.paper?dl=0&rlkey=pr8v3r9e3iclkt5ma87kz8m9n) is not enough?
> 3. Giving enough notice to customers that they cannot upgrade, so that they have enough time to prepare. How much is enough time? 3 days? 1 week? 2 weeks? 1 month?

Yeh, for me points (2) and (3) are the main issues. As an example with what happened with `app-operator`, Adidas had already organised maintenance with team in China, because of timezone differences it was difficult to rearrange that (other organisations might have similar problems with things like CAB). So we want to give the customer sufficient notice so they can organise any internal notifications or changes etc. What sufficient time is, I am not sure... Usually 3days-1 week is enough (some require more, but I wouldn't more than an a week).

Regarding point (1), I know the team have been doing plenty of testing so I am pretty confident everything will be fine. But the best test is always when users are using the service. So we might pick up things shortly after we make the change.

**Comment by cokiengchiara** [comment]
hi @giantswarm/chapter-se 

For Helm 3 Upgrade, we decided:

1. To inspire more confidence, start by upgrading test and safe control planes on Monday
2. Plan to do control plane upgrades on Thursday

On Monday, we're upgrading test CPs and gorilla and gollum. 

Can you give some guidance on which CPs would be safe to upgrade in this [spreadsheet](https://docs.google.com/spreadsheets/d/1ctfe-D6SpnD3mc3FmokttkHfCeDRKVzHk2h8v0KV83o/edit#gid=2025386085)?

**Comment by othylmann** [comment]
General point: This takes 15 minutes, we should be fine telling customers to not upgrade clusters or deploy new ones in that 15 minutes on a per cluster basis. 

In general, known upgrades should be announced a few weeks in advance if at all possible. But I do not really like that happening either because we don't plan that long. Bummer. Just telling you that I feel with you :)

Goal as always: Make it good for the users that do it right, use it to push the customers that are not quite there yet and help those customers get there.

**Comment by rossf7** [comment]
It's been a long time since an update here. So here goes.

## Status

- Main checklist was no longer useful. We now have separate epics for the [CP](https://github.com/giantswarm/giantswarm/issues/10973) and [TC](https://github.com/giantswarm/giantswarm/issues/11748) upgrades.
- We now have separate instances of app-operator for CP and TC. So we can upgrade without impacting customers.
- `giraffe` has been running Helm 3 since 16th June and looks good. 
- CP upgrade plan was merged today.  https://intranet.giantswarm.io/docs/organizational-structure/teams/batman/plans/helm3-control-plane-upgrade/

## TODO

- Upgrading all control planes is scheduled for 30th June to 3rd July.
- #news-product post and ask for help from other teams. Upgrade is automated but needs to be monitored.

**Comment by cokiengchiara** [comment]
Mon, 31. Aug 2020 update
* We are done with upgrading Control Planes to Helm 3
* Next, we are working on platform releases to make Helm 3 available to tenant clusters. We are working on Azure first.

**Comment by cokiengchiara** [comment]
Tue, 8. Sep 2020
* We released in Azure
* Next, we'll release in KVM, then AWS. We're working with a customer to do a preview release on test clusters, to improve safety of upgrades

**Comment by cokiengchiara** [comment]
Done.
